### DISCUSSION TITLE
is 'optional''s implementation  buggy?

### DISCUSSION THREAD

**Discussion by usergoodvery** [protobuf-c/protobuf-c#650] [Q&A] [answered]
Hi,
I spent two days on this now debugging my own app , before I traced it down to the lib's peculiar way of handling optional fields in the protobuf definition file, but I could  be missing something.

Considering the protobuf below, where 'type' is marked as optional, the lib does not pack it even if it has its value set in the source code. Changing that to 'required' packs the value correctly. 

I uploaded a test https://github.com/usergoodvery/protobuf-test to demo that behaviour, based on one of your existing examples. The output from the test code

'type' marked as optional:
$ ./a.out 1 2
Writing 8 serialized bytes
Received: type: '0' a=1 b=2

Change optional to required and regenerate and recompile:
./a.out 1 2
Writing 10 serialized bytes
Received: type: '1' a=1 b=2

message Submessage {
  required int32 value=1;
}

message DMessage {
        enum Type {
        UNKNOWN  = 0;
        REQUEST  = 1;
        RESPONSE = 2;
    }

```
optional Type                     type     = 1;
```

  required Submessage a=2;
  optional Submessage b=3;
}

**Comment by lipnitsk** [comment, marked as answer]
@usergoodvery the problem with your code is that it does not set `msg.has_type` to `TRUE` so the type field is never packed. This is not a bug.

**Comment by usergoodvery** [comment]
Man I wish you documented this requirement somewhere in your examples, would have saved me so much grieve. thanks anyway.
